src/Symfony/Component/Security/Guard/Provider/GuardAuthenticationProvider.php

```diff
@@ -120,8 +127,13 @@ private function authenticateViaGuard(AuthenticatorInterface $guardAuthenticator
 
             throw new BadCredentialsException(sprintf('Authentication failed because "%s::checkCredentials()" did not return true.', get_debug_type($guardAuthenticator)));
         }
-        if ($this->userProvider instanceof PasswordUpgraderInterface && $guardAuthenticator instanceof PasswordAuthenticatedInterface && null !== $this->passwordEncoder && (null !== $password = $guardAuthenticator->getPassword($token->getCredentials())) && method_exists($this->passwordEncoder, 'needsRehash') && $this->passwordEncoder->needsRehash($user)) {
-            $this->userProvider->upgradePassword($user, $this->passwordEncoder->encodePassword($user, $password));
+        if ($this->userProvider instanceof PasswordUpgraderInterface && $guardAuthenticator instanceof PasswordAuthenticatedInterface && null !== $this->passwordHasher && (null !== $password = $guardAuthenticator->getPassword($token->getCredentials())) && method_exists($this->passwordHasher, 'needsRehash') && $this->passwordHasher->needsRehash($user)) {
+            if ($this->passwordHasher instanceof PasswordEncoderInterface) {
+                // @deprecated since Symfony 5.3
+                $this->userProvider->upgradePassword($user, $this->passwordHasher->encodePassword($user, $password));
+            } else {
+                $this->userProvider->upgradePassword($user, $this->passwordHasher->encodePassword($user, $password));
+            }
```

Comment:
`if` and `else` branch execute the same logic. You probably wanted to call the new method in the `else` branch.

Reply:
nice catch! fixed